spec/counter_culture_spec.rb

```diff
@@ -2967,4 +2967,267 @@ def mess_up_counts
+        Transaction.create(monetary_value: 10, person: person)
+      end
+
+      person.reload
+      expect(person.money_earned_total).to eq(10)
+
+      spending_transaction = CounterCulture.aggregate_counter_updates do
+        Transaction.create(monetary_value: -20, person: person)
+      end
+
+      person.reload
+      expect(person.money_spent_total).to eq(-20)
+    end
+
+    it "increments / decrements counter caches correctly for polymorphic association" do
+      require 'models/poly_image'
+      require 'models/poly_employee'
+      require 'models/poly_product'
+
+      employee = PolyEmployee.create(id: 3000)
+      product1 = PolyProduct.create()
+
+      expect(employee.poly_images_count).to eq(0)
+      expect(product1.poly_images_count).to eq(0)
+
+      img1 = CounterCulture.aggregate_counter_updates do
+        PolyImage.create(imageable: employee)
+      end
+
+      expect(employee.reload.poly_images_count).to eq(1)
+      expect(product1.reload.poly_images_count).to eq(0)
+
+      img2 = CounterCulture.aggregate_counter_updates do
+        PolyImage.create(imageable: product1)
+      end
+
+      expect(employee.reload.poly_images_count).to eq(1)
+      expect(product1.reload.poly_images_count).to eq(1)
+
+      img3 = CounterCulture.aggregate_counter_updates do
+        PolyImage.create(imageable: product1)
+      end
+
+      expect(employee.reload.poly_images_count).to eq(1)
+      expect(product1.reload.poly_images_count).to eq(2)
+
+      CounterCulture.aggregate_counter_updates do
+        img3.destroy
+      end
+
+      expect(employee.reload.poly_images_count).to eq(1)
+      expect(product1.reload.poly_images_count).to eq(1)
+
+      CounterCulture.aggregate_counter_updates do
+        img2.imageable = employee
+        img2.save!
+      end
+
+      expect(employee.reload.poly_images_count).to eq(2)
+      expect(product1.reload.poly_images_count).to eq(0)
+    end
+
+    it "works with pg money type" do
+      if ENV['DB'] != 'postgresql'
+        skip("money type only supported in PostgreSQL")
+      end
+
+      item = nil
+      po = PurchaseOrder.create
+
+      expect(po.total_amount).to eq(0.0)
+
+      CounterCulture.aggregate_counter_updates do
+        item = po.purchase_order_items.build(amount: 100.00)
+        item.save
+      end
+
+      po.reload
+      expect(po.total_amount).to eq(100.0)
+
+      CounterCulture.aggregate_counter_updates do
+        item = po.purchase_order_items.build(amount: 100.00)
+        item.save
+      end
+
+      po.reload
+      expect(po.total_amount).to eq(200.0)
+
+      CounterCulture.aggregate_counter_updates do
+        item.destroy
+      end
+
+      po.reload
+      expect(po.total_amount).to eq(100.0)
+
+      CounterCulture.aggregate_counter_updates do
+        po.purchase_order_items.destroy_all
+      end
+
+      po.reload
+      expect(po.total_amount).to eq(0.0)
+    end
+  end
```

Comment:
I think it might be nice to have one "integration-style" test where we do lots of things and make sure that the results end up correct. Maybe some combination of:
* Multiple changes that update the same column, same id
* Other multiple changes that update a different column on the same id, and make that one a `delta_column` rather than a plain counter
* Then some changes on one of the above columns, but for a different id
* A money column change
* A timestamp column change

I have no indication that this would surface anything that won't work, but for aggregation logic like this it's nice to see it all pulled together. And if you combine it with `expect_queries` then we can even make sure that it's aggregating as expected? That last part feels optional, though.

Reply:
I added two more complex tests:
1. to test that counter increments by 0 are ignored
2. a test case that tests:
2.1 Multiple changes that update the same column, same id (both on user and product)
2.2 Other multiple changes that update a different column on the same id, and make that one a delta_column rather than a plain counter (user)
2.3 A timestamp column change (product)
2.4 Then some changes on one of the above columns, but for a different id (you have 2 users and 2 products)

I left the money counter test be on its own (I added a test for it previously)

Reply:
Yeah, this looks great.